test/suites/upgrade.sh

```diff
@@ -0,0 +1,225 @@
+      validate_system_microceph "${m}" 0 0 disk2
+    done
+
+    # Second upgrade MicroOVN.
+    for m in micro01 micro02 micro03; do
+      lxc exec "${m}" -- snap refresh microovn --channel "${microovn_target}"
+    done
+
+    for m in micro01 micro02 micro03; do
+      retries=0
+      while true; do
+        if [ "${retries}" -gt 60 ]; then
+          echo "MicroOVN member ${m} failed to come up after upgrade"
+          exit 1
+        fi
+
+        if lxc exec "${m}" -- sh -c "microovn cluster list -f json | jq -r '.[] | select(.name == \"${m}\") | .status' | grep -q ONLINE"; then
+          break
+        fi
+
+        sleep 1
+        retries=$((retries+1))
+      done
+
+      # There was no explicit OVN underlay subnet configuration in MicroCloud 1.
+      validate_system_microovn "${m}"
+    done
+
+    # Third upgrade LXD.
+    for m in micro01 micro02 micro03; do
+      # Upgrade them in parallel as the refresh waits for the others to update too.
+      lxc exec "${m}" -- snap refresh lxd --channel "${lxd_target}" &
+    done
+    wait
+
+    for m in micro01 micro02 micro03; do
+      retries=0
+      while true; do
+        if [ "${retries}" -gt 60 ]; then
+          echo "LXD member ${m} failed to come up after upgrade"
+          exit 1
+        fi
+
+        if lxc exec "${m}" -- sh -c "lxc cluster list -f json | jq -r '.[] | select(.server_name == \"${m}\") | .status' | grep -q Online"; then
+          break
+        fi
+
+        sleep 1
+        retries=$((retries+1))
+      done
+
+      # Don't test for DNS nameservers on the OVN network as those weren't yet added in MicroCloud 1.
+      validate_system_lxd "${m}" 3 disk1 1 0 enp6s0 10.1.123.1/24 10.1.123.100-10.1.123.254 fd42:1:1234:1234::1/64
+    done
+
+    # Fourth upgrade MicroCloud.
+    for m in micro01 micro02 micro03; do
+      lxc exec "${m}" -- snap refresh microcloud --channel "${microcloud_target}"
+    done
+
+    for m in micro01 micro02 micro03; do
+      retries=0
+      while true; do
+        if [ "${retries}" -gt 60 ]; then
+          echo "LXD member ${m} failed to come up after upgrade"
+          exit 1
+        fi
+
+        if lxc exec "${m}" --env TEST_CONSOLE=0 -- sh -c "microcloud cluster list -f json | jq -r '.[] | select(.name == \"${m}\") | .status' | grep -q ONLINE"; then
+          break
+        fi
+
+        sleep 1
+        retries=$((retries+1))
+      done
+    done
+
+    # Check if the workload survived the upgrade.
+    # Perform some pings to check the network.
+    gateway="$(lxc network get lxdbr0 ipv4.address | awk -F'/' '{print $1}')"
+    if [ "${SKIP_VM_LAUNCH}" = "1" ]; then
+      echo "::warning::SKIPPING VM TESTS"
+    else
+      lxc exec micro01 -- lxc exec v1 -- ping -c 2 "${gateway}"
+
+      # Compare the uptimes to verify the instances haven't been restarted.
+      v1_uptime_after_upgrade="$(lxc exec micro01 -- lxc exec v1 -- uptime -s)"
+      [ "${v1_uptime}" = "${v1_uptime_after_upgrade}" ]
```

Comment:
This `_uptime` variable suffix feels misleading. An alternative would be to compare the `boot_id` which is guarantied to be different between boots. That's what [`lxd-ci` uses](https://github.com/canonical/lxd-ci/blob/main/tests/storage-vm#L686-L697).

Reply:
Thanks, wasn't aware of the boot id. I'll switch to that instead.